crypto/init.c

```diff
@@ -516,14 +514,10 @@ void OPENSSL_cleanup(void)
     err_cleanup();
 
     /* Free list of global locks. */
-    while (global_locks != NULL) {
-        GLOBAL_LOCK *next = global_locks->next;
-
-        free(global_locks);
-        global_locks = next;
+    for (i = 0; i < CRYPTO_NUM_GLOCKS; i++) {
+        CRYPTO_THREAD_lock_free(global_locks[i]);
+        global_locks[i] = NULL;
```

Comment:
This means anyone who calls `fork` after `OPENSSL_cleanup` will UAF since the `pthread_atfork` is still registered.

Reply:
Well, SEGV on the NULL lock, not UAF, but still bad.
Too bad there's no way for OPENSSL_cleanup to unregister the atfork hooks; I guess we'll need to just turn them into noops if the library is not inited, or something like that.

Reply:
How can that happen in the normal case?  The atfork registration would only be called as part of inititialization.

Reply:
Er, yes, SEGV. Although there is a brief window's worth of UAF there if the other thread forks in between the two lines.

@richsalz, I'm thinking an application which has just one random OpenSSL thing to do and afterwards calls `OPENSSL_cleanup`. Something like.
```
  // I only need to compute this one hash on startup. After that, I don't need it,
  // so I'll clean it up.
  OPENSSL_init_crypto(blah);
  EVP_Digest(blah);
  OPENSSL_cleanup();

  // Do the rest of my non-OpenSSL application stuff. Which happens to involve a fork.
  ...
  fork();
  ...
```

In 1.1.0, that works fine. With this `pthread_atfork` lock stuff, it does not since you cannot unregister `pthread_atfork`. If anyone ever `dlclose`s OpenSSL, you will also be sad because the `pthread_atfork` pointers now point to unmapped address.

Whether either of these are supported by OpenSSL, I dunno. `dlclose` is insane and I think not supporting it is quite defensible. I would argue `OPENSSL_cleanup` is similarly absurd, but the function does exist.